### PR TITLE
int64 height (Refs #911)

### PR DESCRIPTION
**Pull request by melekes**

### CODE REVIEW

**Comment by melekes** [comment]
TODO: ensure there is no underflow here

```
consensus/replay.go
115:    gr, found, err = cs.wal.SearchForEndHeight(csHeight - 1)

consensus/state.go
705:    lastBlockMeta := cs.blockStore.LoadBlockMeta(height - 1)

state/execution.go
168:                    signed.SetIndex(i, true) // val_.LastCommitHeight = block.Height - 1

blockchain/reactor.go
58:             store.height-- // XXX HACK, make this better

rpc/core/blocks.go
79:     for height := maxHeight; height >= minHeight; height-- {
```

**Comment by melekes** [comment]
no longer needed
```
consensus/wal_test.go
52:     gr, found, err := wal.SearchForEndHeight(uint64(h))

types/protobuf.go
16:             Height:         uint64(header.Height),

mempool/mempool.go
436:    return uint64(atomic.LoadUint64(&memTx.height))
```

**Comment by zramsay** [comment]
this would close #731

**Comment by ebuchman** [comment]
Needs to be rebased on develop (#835 was merged). Also needs uint64 -> int64

**Comment by melekes** [comment]
```
WARNING: DATA RACE
Read at 0x00c4200bc758 by goroutine 42:
  github.com/tendermint/tendermint/p2p/trust.(*TrustMetric).calcTrustValue()
      github.com/tendermint/tendermint/p2p/trust/_test/_obj_test/trustmetric.go:584 +0xa7
  github.com/tendermint/tendermint/p2p/trust.(*TrustMetric).NextTimeInterval()
      github.com/tendermint/tendermint/p2p/trust/_test/_obj_test/trustmetric.go:348 +0xef
  github.com/tendermint/tendermint/p2p/trust.(*TrustMetric).processRequests()
      github.com/tendermint/tendermint/p2p/trust/_test/_obj_test/trustmetric.go:607 +0xdb

Previous write at 0x00c4200bc758 by goroutine 8:
  [failed to restore the stack]

Goroutine 42 (running) created at:
  github.com/tendermint/tendermint/p2p/trust.NewMetricWithConfig()
      github.com/tendermint/tendermint/p2p/trust/_test/_obj_test/trustmetric.go:448 +0x281
  github.com/tendermint/tendermint/p2p/trust.(*TrustMetricStore).loadFromDB()
      github.com/tendermint/tendermint/p2p/trust/_test/_obj_test/trustmetric.go:147 +0x34a
  github.com/tendermint/tendermint/p2p/trust.TestTrustMetricStoreSaveLoad()
      /go/src/github.com/tendermint/tendermint/p2p/trust/trustmetric_test.go:66 +0x685
  fmt.Fscanf()
      /usr/local/go/src/fmt/scan.go:143 +0xf5
  fmt.Sscanf()
      /usr/local/go/src/fmt/scan.go:114 +0xbe
  github.com/tendermint/tendermint/vendor/github.com/syndtr/goleveldb/leveldb/storage.fsParseName()
      /go/src/github.com/tendermint/tendermint/vendor/github.com/syndtr/goleveldb/leveldb/storage/file_storage.go:571 +0x18c
  github.com/tendermint/tendermint/vendor/github.com/syndtr/goleveldb/leveldb/storage.(*fileStorage).List()
      /go/src/github.com/tendermint/tendermint/vendor/github.com/syndtr/goleveldb/leveldb/storage/file_storage.go:386 +0x2dd
```

I thought we'd fixed that.

**Comment by ebuchman** [comment]
> I thought we'd fixed that.

Looks like a different race :(

File: mempool/mempool.go [hunk 1]
```diff
@@ -432,8 +432,8 @@ type mempoolTx struct {
 }
 
 // Height returns the height for this transaction
-func (memTx *mempoolTx) Height() int {
-	return int(atomic.LoadInt64(&memTx.height))
+func (memTx *mempoolTx) Height() int64 {
+	return atomic.LoadInt64(&memTx.height)
```
Comment (ebuchman) [inline mempool/mempool.go]: heh!

File: rpc/client/helpers.go [hunk 2]
```diff
@@ -42,7 +42,7 @@ func WaitForHeight(c StatusClient, h int, waiter Waiter) error {
 		if err != nil {
 			return err
 		}
-		delta = h - s.LatestBlockHeight
+		delta = int(h - s.LatestBlockHeight)
```
Comment (ebuchman) [inline rpc/client/helpers.go]: ?

File: rpc/core/blocks.go [hunk 3]
```diff
@@ -61,20 +61,27 @@ import (
...
+		maxHeight = cmn.MinInt64(blockStore.Height(), maxHeight)
 	}
 
+	// maximum 20 block metas
+	const limit int64 = 20
+	minHeight = cmn.MaxInt64(minHeight, maxHeight-limit)
+
 	logger.Debug("BlockchainInfoHandler", "maxHeight", maxHeight, "minHeight", minHeight)
 
+	if minHeight > maxHeight {
```
Comment (ebuchman) [inline rpc/core/blocks.go]: nice thanks

File: rpc/core/tx.go [hunk 4]
```diff
@@ -84,18 +84,18 @@ func Tx(hash []byte, prove bool) (*ctypes.ResultTx, error) {
 	}
 
 	height := r.Height
-	index := r.Index
+	index := int(r.Index) // XXX:overflow
```
Comment (ebuchman) [inline rpc/core/tx.go]: cool should be fixed with new abci

File: rpc/core/tx.go [hunk 5]
```diff
@@ -187,8 +187,8 @@ func TxSearch(query string, prove bool) ([]*ctypes.ResultTx, error) {
 		index := r.Index
 
 		if prove {
+			block := blockStore.LoadBlock(height)
 			// TODO: handle overflow
-			block := blockStore.LoadBlock(int(height))
 			proof = block.Data.Txs.Proof(int(index))
```
Comment (ebuchman) [inline rpc/core/tx.go]: can remove this `int`. but lets just wait till abci update

File: types/validator_set.go [hunk 6]
```diff
@@ -53,7 +53,7 @@ func (valSet *ValidatorSet) IncrementAccum(times int) {
 	// Add VotingPower * times to each validator and order into heap.
 	validatorsHeap := cmn.NewHeap()
 	for _, val := range valSet.Validators {
-		val.Accum += int64(val.VotingPower) * int64(times) // TODO: mind overflow
+		val.Accum += val.VotingPower * int64(times) // TODO: mind overflow
```
Comment (ebuchman) [inline types/validator_set.go]: this one's actually a little scary.

File: consensus/replay.go [hunk 7]
```diff
@@ -90,15 +90,16 @@ func (cs *ConsensusState) readReplayMessage(msg *TimedWALMessage, newStepCh chan
 
 // replay only those messages since the last block.
 // timeoutRoutine should run concurrently to read off tickChan
-func (cs *ConsensusState) catchupReplay(csHeight int) error {
+// CONTRACT: csHeight > 0
```
[the remark this answers is not in the document]
Comment (melekes) [reply consensus/replay.go]: this was for uint. i'll remove

File: rpc/client/helpers.go (hunk 2 above)
Comment (melekes) [reply rpc/client/helpers.go]: latestblockheight is int64. delta is int

File: rpc/core/tx.go (hunk 5 above)
Comment (melekes) [reply rpc/core/tx.go]: should linter catch such things?

File: types/validator_set.go (hunk 6 above)
Comment (melekes) [reply types/validator_set.go]: very scary

File: rpc/client/helpers.go (hunk 2 above)
Comment (ebuchman) [reply rpc/client/helpers.go]: shouldnt we change delta to int64?

File: rpc/core/tx.go (hunk 5 above)
Comment (ebuchman) [reply rpc/core/tx.go]: not sure @zramsay . we dont have all linters active yet

File: types/validator_set.go (hunk 6 above)
Comment (ebuchman) [reply types/validator_set.go]: opened an issue https://github.com/tendermint/tendermint/issues/919

File: rpc/client/helpers.go (hunk 2 above)
Comment (melekes) [reply rpc/client/helpers.go]: yes!

**Comment by silasdavis** [comment]
@melekes missed this when it was going through. Any reason why we are not using `uint64` since block height cannot be negative. I have been using `uint64` on my side and was hoping that you would adopt that too...

**Comment by ebuchman** [comment]
uints are bad for arithmetic. After extensive research, the general consensus seems to be that uints are only good for things that will never be involved in arithmetic - eg. bitarrays/masks.

The risk of messing up the arithmetic is much greater than the type enforcement that it will not be negative.

**Pull request merged**